test/jdk/java/net/httpclient/websocket/java.net.http/jdk/internal/net/http/websocket/WebSocketAndHttpClient.java

```diff
@@ -0,0 +1,114 @@
+package jdk.internal.net.http.websocket;
+
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.net.http.WebSocket;
+import java.util.Optional;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+
+/**
+ * This is the client side of the test invoked from WebSocketAndHttpTest:
+ *
+ * The two args are the addresses of a (local) Websocket and Http server
+ *
+ * The test first sends a request to the WS server and in the listener
+ * which handles the response, it tries to send a request to the http
+ * server. This hangs if the listener was invoked from the selector
+ * manager thread. If invoked from a different thread then the http
+ * response is received and the response string is mapped to string
+ * "succeeded"
+ */
+public class WebSocketAndHttpClient {
+
+    public static void main(String[] args) throws InterruptedException {
+
+        ExecutorService executorService = Executors.newCachedThreadPool();
+        HttpClient httpClient = HttpClient.newBuilder().executor(executorService).build();
+
+        WebSocketTest wsTest = new WebSocketTest(httpClient, args[0]);
+        HttpTest httpTest = new HttpTest(httpClient, args[1]);
+
+        AtomicReference<String> result = new AtomicReference<>("failed");
```

Comment:
Would it be possible to use a CountDownLatch - or a CompletableFuture<String> instead of an atomic reference in order to replace the Thread.sleep(3_000) at line 44 with latch.await() or cf.join()?
The test would then fail in jtreg timeout without the fix, but would not have to wait for an arbitrary magic 3s delay if the fix is present (which might not be enough delay on slow machines with debug builds etc...)

Reply:
Good idea on the CF or latch. As regards the blind cast, I suppose I could test for it and if the type is different, leave the executor reference as null, which reverts current behavior, but I can't imagine a scenario where that would actually happen. If it did, would we want to revert the behavior, or fail visibly with CCE?

Reply:
I agree - I can't see how it could be something else than HttpClientFacade with the current implementation. Maybe just let's leave that up for later cleanup. What I had in mind was adding a static helper method somewhere on a public class in `jdk.internal.net.http` that would have access to the package private APIs - something like:

```
public static Executor getClientExecutor(HttpClient client) {
   ... and here we could do some switch depending on the concrete subclass of HttpClient
       we're dealing with ...
}
```

But I don't see an obvious home for such a method - so maybe we should leave this for a later cleanup.